tools/upload-grid.mjs

```diff
@@ -0,0 +1,83 @@
+import fs from 'fs';
+import { getEnvKey } from './lib/envkeys.mjs';
+import { fetchProject, assignSessionsToSlotAndRoom } from './lib/project.mjs'
+import { validateSession } from './lib/validate.mjs';
+import seedrandom from 'seedrandom';
+
+function readconfig({ filename }) {
+   if (filename) {
+      let content = fs.readFileSync(filename).toString();
+      let regexp = /<pre id="data">[^<]+<\/pre>/s;
+      let data = content.match(regexp)[0].match(/\[\{.*\}\]/)[0];
+      return(Array.from(JSON.parse(data)));
+      }
+   }
+
+async function main({ filename, apply }) {
+  const PROJECT_OWNER = await getEnvKey('PROJECT_OWNER');
+  const PROJECT_NUMBER = await getEnvKey('PROJECT_NUMBER');
+  console.warn();
+  console.warn(`Retrieve project ${PROJECT_OWNER}/${PROJECT_NUMBER}...`);
+  const project = await fetchProject(PROJECT_OWNER, PROJECT_NUMBER);
+  if (!project) {
+    throw new Error(`Project ${PROJECT_OWNER}/${PROJECT_NUMBER} could not be retrieved`);
+  }
+  console.warn(`- found ${project.sessions.length} sessions`);
+  let sessions = await Promise.all(project.sessions);
+  sessions = sessions.filter(s => !!s);
+  console.warn(`Retrieve project ${PROJECT_OWNER}/${PROJECT_NUMBER} and session(s)... done`);
+
+  const rooms = project.rooms;
+  const slots = project.slots;
+  const configs = readconfig({ filename });
+    
+  for (const config of configs) {
+      if (!sessions.map(s => s.number).includes(config.number)) {
+     	  console.warn("Unknown session " + config.number);
+//	  continue;
```

Comment:
If session is unknown, if room is unknown or if slot is unknown, the assignSessionsToSlotAndRoom function will fail as the GitHub API endpoint can only save data it understands!